Домашнее задание

anv0l opened the pull request:
Автоматическое логирование.
Цель: Понять как реализуется AOP, какие для этого есть технические средства.
Разработайте такой функционал:
метод класса можно пометить самодельной аннотацией @Log, например, так:

class TestLogging {
@Log
public void calculation(int param) {};
}

При вызове этого метода "автомагически" в консоль должны логироваться значения параметров.
Например так.

class Demo {
public void action() {
new TestLogging().calculation(6);
}
}

В консоле дожно быть:
executed method: calculation, param: 6

Обратите внимание: явного вызова логирования быть не должно.

anv0l committed:
Домашнее задание

Автоматическое логирование.
Цель: Понять как реализуется AOP, какие для этого есть технические средства.
Разработайте такой функционал:
метод класса можно пометить самодельной аннотацией @Log, например, так:

class TestLogging {
@Log
public void calculation(int param) {};
}

При вызове этого метода "автомагически" в консоль должны логироваться значения параметров.
Например так.

class Demo {
public void action() {
new TestLogging().calculation(6);
}
}

В консоле дожно быть:
executed method: calculation, param: 6

Обратите внимание: явного вызова логирования быть не должно.

aimiebell commented on hw05-aop/src/main/java/homework/Calculations.java:
```diff
@@ -0,0 +1,13 @@
+
+public class Calculations {
+    public static void main(String[] args) {
+        Calculator calculator = Logger.CreateCalculator();
```
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

aimiebell replied:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

aimiebell commented on hw05-aop/src/main/java/homework/Logger.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    // не совсем понял, что тут происходит, просто скопировал код
+    static Calculator CreateCalculator() {
```
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

aimiebell replied:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

anv0l committed:
CreateCalculator -> createCalculator

petrelevich commented on hw05-aop/src/main/java/homework/Log.java:
```diff
@@ -0,0 +1,8 @@
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
```
не забывайте про target

anv0l replied:
Добавил Target.

petrelevich commented on hw05-aop/src/main/java/homework/Calculator.java:
```diff
@@ -0,0 +1,14 @@
+package homework;
+
+public interface Calculator {
+    @Log
```
аннотация должна быть на методах класса, а не интерфейса.
это важно.

anv0l replied:
Перенёс аннотацию на методы класса.

petrelevich commented on hw05-aop/src/main/java/homework/Logger.java:
```diff
@@ -0,0 +1,41 @@
+import java.lang.reflect.Proxy;
+import java.util.Arrays;
+
+public class Logger {
```
В этом классе все же не функционал логирования, а функционал создания экземпляра объекта.

anv0l replied:
Переименовал.

petrelevich commented on hw05-aop/src/main/java/homework/Logger.java:
```diff
@@ -0,0 +1,41 @@
+    private Logger() {
+    }
+
+    // не совсем понял, что тут происходит, просто скопировал код
```
Proxy.newProxyInstance создает прокси-объект, с имплементацией интерфейса Calculator.
логика работы прокси реализуется в handler

petrelevich commented on hw05-aop/src/main/java/homework/Logger.java:
```diff
@@ -0,0 +1,41 @@
+            Annotation[] annotations = method.getDeclaredAnnotations();
+            String packageName = this.getClass().getPackageName();
+            packageName = packageName.isBlank() ?"":packageName + ".";
+            if (Arrays.asList(annotations).toString().contains( packageName + LOGGER_ANNOTATION)) {
```
посмотрите метод
getClass().isAnnotationPresent()

поиск методов, на которых стоит аннотация, надо сделать один раз при создании InvocationHandler.
В log надо использовать результаты этого анализа.

anv0l replied:
Перенёс получение списка метода с аннотацией @Log во время создания InvocationHandler, теперь проверяем только в этом списке, а не получаем список всех аннотаций для всех методов при каждом вызове.

petrelevich commented on hw05-aop/src/main/java/homework/MyCalculator.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public long sum(long a, long b) {
+        return a + b;
+    }
```
хорошо бы проверить перегрузку методов:
```
public void calculation(int param1)
public void calculation(int param1, int param2)
public void calculation(int param1, int param2, String param3)
```
как с ними будет работать.

anv0l replied:
В новой версии работает.

anv0l committed:
1. Изменил название класса, в котором создаётся прокси-объект.
2. При создании прокси-объекта также получаем список методов с аннотацией @Log.
3. При логгировании вызываемый метод справнивается с методами из списка, а не вообще со всеми класса.
4. Изменил название методов в Calculator на более абстрактные.

petrelevich commented on hw05-aop/src/main/java/homework/CalculatorProxy.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.stream.Collectors;
+
+public class CalculatorProxy {
+    private static final Class LOGGER_ANNOTATION = Log.class;
```
лучше так:
`private static final Class<? extends Annotation> LOGGER_ANNOTATION = Log.class;`

anv0l replied:
Изменил.

petrelevich commented on hw05-aop/src/main/java/homework/CalculatorProxy.java:
```diff
@@ -0,0 +1,46 @@
+        Method[] methods = MyCalculator.class.getMethods();
+        logMethods = Arrays.stream(methods)
+                .filter(method -> method.isAnnotationPresent(LOGGER_ANNOTATION))
+                .map(method -> method.getName() + Arrays.asList(method.getParameterTypes()).toString())
```
"Сворачиваение метода" к строковому идентификатору лучше в отдельный метод вынести.

anv0l replied:
Вынес метод, речь ведь идёт только про получение названия и параметра одного метода?

petrelevich replied:
да, и этот же метод надо использовать при проверке наличия аннотации.

petrelevich commented on hw05-aop/src/main/java/homework/CalculatorProxy.java:
```diff
@@ -0,0 +1,46 @@
+
+        private void log(Method method, Object[] args) {
+            String methodParams = method.getName() + Arrays.asList(method.getParameterTypes()).toString();
+            if (logMethods.contains(methodParams)) {
```
logMethods - лучше сделать HashSet, чтобы contains работал быстрее.

anv0l replied:
Заменил.

anv0l committed:
1. Тип аннотации логгирования сделал Class<? extends Annotation>.
2. Вынес получение названия и параметров метода в отдельный метод.
3. Методы теперь сохраняются в HashSet.
